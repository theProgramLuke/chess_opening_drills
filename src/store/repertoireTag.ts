import { Side } from "./side";

export class RepertoireTag {
  id: number;
  forSide: Side;
  name: string;
  position: string;
  children: Array<RepertoireTag>;

  constructor(
    id: number,
    forSide: Side,
    name: string,
    position: string,
    children: Array<RepertoireTag>
  ) {
    this.id = id;
    this.forSide = forSide;
    this.name = name;
    this.position = position;
    this.children = children;
  }
}

function _findTag(tag: RepertoireTag, id: number): RepertoireTag | null {
  if (tag["id"] === id) {
    return tag;
  }

  if (tag["children"].length === 0) {
    return null;
  }

  const mapped = tag["children"].map((child: RepertoireTag) =>
    _findTag(child, id)
  );
  const filtered = mapped.filter((child: RepertoireTag | null) => {
    return child != null;
  });

  return filtered[0];
}

export function FindRepertoireTag(
  tags: Array<RepertoireTag>,
  id: number
): RepertoireTag | null {
  const mapped = tags.map((child: RepertoireTag) => _findTag(child, id));
  const filtered = mapped.filter((child: RepertoireTag | null) => {
    return child != null;
  });

  // assert(filtered.length === 1, "Should only find one matching tag...");

  return filtered[0];
}
