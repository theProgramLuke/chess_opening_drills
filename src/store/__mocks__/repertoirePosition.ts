import _ from "lodash";

import { Move } from "../move";
import { Side } from "../side";
import { TrainingEvent } from "../TrainingEvent";
import { TrainingMode } from "../trainingMode";
import { Turn } from "../turn";
import { FEN } from "chessground/types";
import { SavedRepertoirePosition } from "../repertoirePosition";

export class RepertoirePosition {
  fen: string;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;
  trainingHistory: TrainingEvent[];
  nextRepetitionTimestamp?: number | undefined;
  previousIntervalDays: number;
  intervalIndex: number;
  easinessFactor: number;
  SideToMove: () => Side;
  AddChild: (move: Move) => void;
  Unlink: () => void;
  IncludeForTrainingMode: (
    mode: TrainingMode,
    difficultyModeLimit?: number
  ) => boolean;
  AddTrainingEvent: (event: TrainingEvent) => void;
  VisitChildren: (visit: (child: RepertoirePosition) => void) => void;
  IsChildPosition: (maybeChild: RepertoirePosition) => boolean;
  RootPaths: () => Move[][];
  GetTurnLists: () => Turn[][];
  AsPgn: () => string;
  AsPgnMoveText: (turnCount?: number) => string;
  AsSaved: (childParentSource: RepertoirePosition[]) => SavedRepertoirePosition;

  constructor(fen: FEN, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
    this.myTurn = false;
    this.forSide = Side.White;
    this.trainingHistory = [];
    this.nextRepetitionTimestamp = 0;
    this.previousIntervalDays = 0;
    this.intervalIndex = 0;
    this.easinessFactor = 0;
    this.SideToMove = jest.fn();
    this.AddChild = jest.fn();
    this.Unlink = jest.fn();
    this.IncludeForTrainingMode = jest.fn();
    this.AddTrainingEvent = jest.fn();
    this.VisitChildren = jest.fn();
    this.IsChildPosition = jest.fn();
    this.RootPaths = jest.fn();
    this.GetTurnLists = jest.fn();
    this.AsPgn = jest.fn();
    this.AsPgnMoveText = jest.fn();
    this.AsSaved = jest.fn();
  }
}
