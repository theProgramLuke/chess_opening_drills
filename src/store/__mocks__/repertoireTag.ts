import { FEN } from "chessground/types";

import * as Real from "@/store/repertoireTag";
import { Side } from "@/store/side";
import { RepertoirePosition } from "@/store/repertoirePosition";

export class RepertoireTag {
  forSide: Side;
  name: string;
  position: RepertoirePosition;
  fen: FEN;
  children: RepertoireTag[];
  id: string;

  constructor(
    forSide: Side,
    name: string,
    position: RepertoirePosition,
    fen: FEN,
    children: RepertoireTag[],
    id?: string
  ) {
    this.id = id || "";
    this.forSide = forSide;
    this.name = name;
    this.position = position;
    this.fen = fen;
    this.children = children;
    this.AddChild = jest.fn();
    this.RemoveChild = jest.fn();
    this.AsSaved = jest.fn();
  }

  AddChild: (tag: RepertoireTag) => void;

  RemoveChild: (tag: RepertoireTag) => void;

  AsSaved: (positionSource: RepertoirePosition[]) => Real.SavedRepertoireTag;

  static FromSaved: (
    saved: Real.SavedRepertoireTag,
    positionSource: RepertoirePosition[]
  ) => RepertoireTag = jest.fn();
}

export const GetTrainingPositions = jest.fn();

export const GetTrainingMoveLists = jest.fn();
