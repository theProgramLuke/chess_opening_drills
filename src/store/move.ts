import { RepertoirePosition } from "@/store/repertoirePosition";

export class Move {
  san: string;
  position: RepertoirePosition;

  constructor(san: string, position: RepertoirePosition) {
    this.san = san;
    this.position = position;
  }
}
