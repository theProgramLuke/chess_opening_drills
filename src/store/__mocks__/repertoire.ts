import _ from "lodash";

import {
  RepertoirePosition,
  SavedRepertoirePosition
} from "@/store/repertoirePosition";
import { RepertoireTag, SavedRepertoireTag } from "@/store/repertoireTag";
import { Move } from "@/store/move";
import { PgnGame } from "pgn-parser";

export class SavedRepertoire {
  positions: SavedRepertoirePosition[];
  tags: SavedRepertoireTag[];

  constructor(
    positions: SavedRepertoirePosition[],
    tags: SavedRepertoireTag[]
  ) {
    this.positions = positions;
    this.tags = tags;
  }
}

export class Repertoire {
  positions: RepertoirePosition[];
  tags: RepertoireTag[];

  constructor(positions: RepertoirePosition[], tags: RepertoireTag[]) {
    this.positions = positions;
    this.tags = tags;
    this.AddMove = jest.fn();
    this.RemoveRepertoireTag = jest.fn();
    this.RemoveMove = jest.fn();
    this.AddFromGame = jest.fn();
    this.AsSaved = jest.fn();
  }

  AddMove: (parent: RepertoirePosition, move: Move) => void;

  RemoveRepertoireTag: (tagToRemove: RepertoireTag) => void;

  RemoveMove: (move: Move) => void;

  AddFromGame: (game: PgnGame) => void;

  AsSaved: () => SavedRepertoire;

  static FromSaved: (saved: SavedRepertoire) => Repertoire = jest.fn();
}
