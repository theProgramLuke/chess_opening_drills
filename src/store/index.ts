import Vue from "vue";
import Vuex from "vuex";

import { RepertoirePosition } from "@/store/repertoirePosition";
import { Move } from "./move";
import { GetPersistantStorage } from "./storage";
import { Repertoire } from "./repertoire";
import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { TrainingEvent } from "./TrainingEvent";
import { PgnGame } from "./pgnParser";

Vue.use(Vuex);

const storage = GetPersistantStorage();

export default new Vuex.Store({
  state: {
    darkMode: storage.get("darkMode"),
    primary: storage.get("primary"),
    secondary: storage.get("secondary"),
    accent: storage.get("accent"),
    error: storage.get("error"),
    warning: storage.get("warning"),
    info: storage.get("info"),
    success: storage.get("success"),
    boardTheme: storage.get("boardTheme"),
    pieceTheme: storage.get("pieceTheme"),
    whiteRepertoire: Repertoire.FromSaved(storage.get("whiteRepertoire")),
    blackRepertoire: Repertoire.FromSaved(storage.get("blackRepertoire"))
  },

  mutations: {
    setDarkMode: (state, darkMode): void => {
      state.darkMode = darkMode;
      storage.set("darkMode", darkMode);
    },

    setBoardTheme: (state, boardTheme): void => {
      state.boardTheme = boardTheme;
      storage.set("boardTheme", boardTheme);
    },

    setPieceTheme: (state, pieceTheme): void => {
      state.pieceTheme = pieceTheme;
      storage.set("pieceTheme", pieceTheme);
    },

    setColor: function(
      state,
      payload: {
        colorToSet:
          | "primary"
          | "secondary"
          | "accent"
          | "error"
          | "warning"
          | "info"
          | "success";
        color: string;
      }
    ): void {
      state[payload.colorToSet] = payload.color;
      storage.set(payload.colorToSet, payload.color);
    },

    addRepertoirePosition: (
      state,
      payload: { parent: RepertoirePosition; newMove: Move }
    ): void => {
      if (payload.parent && payload.newMove) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddMove(payload.parent, payload.newMove);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addRepertoireTag: (
      state,
      payload: { parent: RepertoireTag; tag: RepertoireTag }
    ): void => {
      if (payload.parent && payload.tag) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.parent.AddChild(payload.tag);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addTrainingEvent: (
      state,
      payload: { position: RepertoirePosition; event: TrainingEvent }
    ): void => {
      if (payload.position && payload.event) {
        const repertoire =
          payload.position.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.position.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.position.AddTrainingEvent(payload.event);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    removeRepertoireTag: (state, tag: RepertoireTag): void => {
      const repertoire =
        tag.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        tag.forSide === Side.White ? "whiteRepertoire" : "blackRepertoire";

      repertoire.RemoveRepertoireTag(tag);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    removeRepertoireMove: (state, move: Move): void => {
      const repertoire =
        move.position.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        move.position.forSide === Side.White
          ? "whiteRepertoire"
          : "blackRepertoire";

      repertoire.RemoveMove(move);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    addPositionsFromGame(
      state,
      payload: { forSide: Side; game: PgnGame }
    ): void {
      if (payload.game) {
        const repertoire =
          payload.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddFromGame(payload.game);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    clearStorage(): void {
      storage.clear();
    }
  },

  actions: {},

  modules: {},

  strict: process.env.NODE_ENV !== "production"
});
