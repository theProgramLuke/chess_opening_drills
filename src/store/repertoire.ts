import _ from "lodash";
import { FEN } from "chessground/types";

import {
  RepertoirePosition,
  SavedRepertoirePosition
} from "@/store/repertoirePosition";
import { RepertoireTag, SavedRepertoireTag } from "@/store/repertoireTag";
import { Move } from "@/store/move";

export class SavedRepertoire {
  positions: SavedRepertoirePosition[];
  tags: SavedRepertoireTag[];

  constructor(
    positions: SavedRepertoirePosition[],
    tags: SavedRepertoireTag[]
  ) {
    this.positions = positions;
    this.tags = tags;
  }
}

export class Repertoire {
  positions: RepertoirePosition[];
  tags: RepertoireTag[];

  constructor(positions: RepertoirePosition[], tags: RepertoireTag[]) {
    this.positions = positions;
    this.tags = tags;
  }

  AddMove(parent: RepertoirePosition, move: Move): void {
    const existingPosition = this.LookupRepertoirePosition(move.position.fen);
    if (existingPosition) {
      move.position = existingPosition;
    }

    parent.AddChild(move);

    if (!existingPosition) {
      this.positions.push(move.position);
    }
  }

  private LookupRepertoirePosition(fen: FEN): RepertoirePosition | undefined {
    return _.find(this.positions, (position: RepertoirePosition) => {
      return fen === position.fen;
    });
  }

  AsSaved(): SavedRepertoire {
    const savedPositions = _.map(this.positions, position =>
      position.AsSaved(this.positions)
    );

    const loadedTags = _.map(this.tags, tag => tag.AsSaved(this.positions));

    return new SavedRepertoire(savedPositions, loadedTags);
  }

  static FromSaved(saved: SavedRepertoire): Repertoire {
    const loadedPositions = _.map(saved.positions, savedPosition =>
      RepertoirePosition.FromSaved(savedPosition)
    );

    _.forEach(saved.positions, (savedPosition, index) =>
      _.forEach(savedPosition.children, savedMove =>
        loadedPositions[index].AddChild(
          new Move(savedMove.san, loadedPositions[savedMove.positionId])
        )
      )
    );

    const loadedTags = _.map(saved.tags, savedTag =>
      RepertoireTag.FromSaved(savedTag, loadedPositions)
    );

    return new Repertoire(loadedPositions, loadedTags);
  }
}
