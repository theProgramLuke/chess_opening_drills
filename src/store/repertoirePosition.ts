import _ from "lodash";
import { FEN } from "chessground/types";
import { Chess } from "chess.js";

import { Move, SavedMove } from "@/store/move";
import { Side } from "@/store/side";
import { Turn } from "@/store/turn";

export class SavedRepertoirePosition {
  fen: FEN;
  comment: string;
  parentIds: number[];
  children: SavedMove[];
  myTurn: boolean;
  forSide: Side;

  constructor(
    fen: FEN,
    comment: string,
    parentIds: number[],
    children: SavedMove[],
    myTurn: boolean,
    forSide: Side
  ) {
    this.fen = fen;
    this.comment = comment;
    this.parentIds = parentIds;
    this.children = children;
    this.myTurn = myTurn;
    this.forSide = forSide;
  }
}

export class RepertoirePosition {
  fen: FEN;
  comment: string;
  parents: RepertoirePosition[];
  children: Move[];
  myTurn: boolean;
  forSide: Side;

  constructor(fen: FEN, myTurn: boolean, comment: string, forSide: Side) {
    this.fen = fen;
    this.comment = comment;
    this.myTurn = myTurn;
    this.forSide = forSide;
    this.parents = [];
    this.children = [];
  }

  SideToMove(): Side {
    return Chess(this.fen).turn() === "w" ? Side.White : Side.Black;
  }

  AddChild(move: Move): void {
    const samePosition = this.fen == move.position.fen;
    const alreadyChild = !_.isEmpty(
      _.find(
        this.children,
        (child: Move) => move.position.fen === child.position.fen
      )
    );
    if (!samePosition && !alreadyChild) {
      this.children.push(move);
      move.position.parents.push(this);
    }
  }

  private RootPaths(): Move[][] {
    const collector: Move[][] = [];

    if (!_.isEmpty(this.parents)) {
      this.RootPathsRecursive([], collector);
    }

    return collector;
  }

  private RootPathsRecursive(path: Move[], collector: Move[][]): void {
    if (_.isEmpty(this.parents)) {
      collector.push(path);
    } else {
      _.forEach(this.parents, parent => {
        const branch = _.clone(path);
        const parentMove = parent.getChildMove(this);

        if (parentMove) {
          branch.unshift(parentMove);
        }

        parent.RootPathsRecursive(branch, collector);
      });
    }
  }

  private getChildMove(other: RepertoirePosition): Move | undefined {
    return _.find(this.children, child => child.position === other);
  }

  GetTurnLists(): Array<Array<Turn>> {
    const moveLists: Array<Array<Turn>> = [];

    const paths = this.RootPaths();

    _.forEach(paths, (path: Array<Move>) => {
      const turns: Array<Turn> = [];

      _.forEach(_.range(0, path.length, 2), (i: number) => {
        const whiteMove = path[i];

        if (path.length === i) {
          turns.push(new Turn(whiteMove));
        } else {
          const blackMove = path[i + 1];
          turns.push(new Turn(whiteMove, blackMove));
        }
      });

      moveLists.push(turns);
    });

    return moveLists;
  }

  AsPgnMoveText(turnCount = 1): string {
    if (_.isEmpty(this.children)) {
      return "";
    }

    const side = this.SideToMove();
    let pgnMoveText = "";
    let childTurnCount = turnCount;

    if (side === Side.White) {
      // White made this move.
      pgnMoveText += turnCount + ". ";
      childTurnCount++;
    }

    const mainLine = _.first(this.children);
    if (mainLine) {
      pgnMoveText += mainLine.san;
    }

    const variations = _.tail(this.children);

    _.forEach(variations, child => {
      pgnMoveText += " ( "; // start variation
      pgnMoveText += turnCount;
      pgnMoveText += side === Side.White ? ". " : "... ";
      pgnMoveText += child.san;
      pgnMoveText += " ";
      pgnMoveText += child.position.AsPgnMoveText(childTurnCount);
      pgnMoveText += " )"; // end variation
    });

    if (mainLine) {
      pgnMoveText += " ";
      pgnMoveText += mainLine.position.AsPgnMoveText(childTurnCount);
    }

    return _.trim(pgnMoveText);
  }

  AsSaved(childParentSource: RepertoirePosition[]): SavedRepertoirePosition {
    const savedChildren = _.map(
      this.children,
      child =>
        new SavedMove(child.san, _.indexOf(childParentSource, child.position))
    );

    const parentIds = _.map(this.parents, parent =>
      _.indexOf(childParentSource, parent)
    );

    return new SavedRepertoirePosition(
      this.fen,
      this.comment,
      parentIds,
      savedChildren,
      this.myTurn,
      this.forSide
    );
  }

  static FromSaved(saved: SavedRepertoirePosition): RepertoirePosition {
    return new RepertoirePosition(
      saved.fen,
      saved.myTurn,
      saved.comment,
      saved.forSide
    );
  }
}
