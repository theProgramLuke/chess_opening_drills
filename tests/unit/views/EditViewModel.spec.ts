import { shallowMount, createLocalVue } from "@vue/test-utils";
import _ from "lodash";
import Vuex, { Store } from "vuex";

import EditViewModel from "@/views/EditViewModel.ts";
import { Repertoire } from "@/store/repertoire";
import { RepertoirePosition } from "@/store/repertoirePosition";
import { Side } from "@/store/side";
import { RepertoireTag } from "@/store/repertoireTag";
import { Turn } from "@/store/turn";
import { Move } from "@/store/move";

jest.mock("@/store/repertoire");
jest.mock("@/store/repertoireTag");
jest.mock("@/store/repertoirePosition");

const mutations = {
  addRepertoirePosition: jest.fn(),
  addRepertoireTag: jest.fn()
};
let startPosition: RepertoirePosition;
let state: { whiteRepertoire: Repertoire; blackRepertoire: Repertoire };
const localVue = createLocalVue();
localVue.use(Vuex);
let store: Store<unknown>;

beforeEach(() => {
  startPosition = new RepertoirePosition("", "", Side.White);
  state = {
    whiteRepertoire: new Repertoire(
      [],
      [new RepertoireTag(Side.White, "", startPosition, "", [])]
    ),
    blackRepertoire: new Repertoire([], [])
  };

  store = new Vuex.Store({ state, mutations });
});

const makeTurn = (count = 0) =>
  new Turn(
    new Move(count.toString(), new RepertoirePosition("", "", Side.White))
  );

describe("EditViewModel", () => {
  const mountComponent = () =>
    shallowMount(EditViewModel, {
      render: jest.fn(),
      localVue,
      store
    });

  beforeEach(() => {
    _.forEach(mutations, mutation => mutation.mockReset());
  });

  describe("created", () => {
    it("should be created with the white repertoire start position", () => {
      const component = mountComponent();

      const activePosition = component.vm.activePosition;

      expect(component.vm.activePosition).toBe(startPosition);
    });
  });

  describe("turnLists", () => {
    it("should be the turn lists of the active position", () => {
      const turnList = [_.times(1, makeTurn), _.times(5, makeTurn)];
      (startPosition.GetTurnLists as jest.Mock).mockReturnValue(turnList);
      const component = mountComponent();

      const actual = component.vm.turnLists;

      expect(actual).toBe(turnList);
    });

    it("should be [[]] if there are no moves", () => {
      (startPosition.GetTurnLists as jest.Mock).mockReturnValue(false);
      const component = mountComponent();

      const actual = component.vm.turnLists;

      expect(actual).toEqual([[]]);
    });
  });

  describe("nextMoves", () => {
    it("should be the next moves of the active position", () => {
      startPosition.children = [
        new Move("0", new RepertoirePosition("", "", Side.White)),
        new Move("1", new RepertoirePosition("", "", Side.White))
      ];
      const component = mountComponent();

      const actual = component.vm.nextMoves;

      expect(actual).toEqual(startPosition.children);
    });
  });

  describe("updateBoard", () => {
    it("should update the active position and board orientation", () => {
      const side = Side.Black;
      const component = mountComponent();
      const updatedPosition = new RepertoirePosition("some fen", "", side);

      component.vm.updateBoard(updatedPosition);

      expect(component.vm.boardOrientation).toEqual(side);
      expect(component.vm.activePosition).toBe(updatedPosition);
    });
  });

  describe("onBoardMove", () => {
    it("should add a repertoire position with the new move", () => {
      const component = mountComponent();
      const fen = "new move fen";
      const san = "e5";
      const history = ["e4", san];

      component.vm.onBoardMove({ fen, history });

      expect(mutations.addRepertoirePosition).toMatchSnapshot();
    });

    it("should update the board with the new move", () => {
      const component = mountComponent();
      component.vm.updateBoard = jest.fn();

      component.vm.onBoardMove({ fen: "fen" });

      expect(component.vm.updateBoard).toBeCalledWith(
        expect.any(RepertoirePosition)
      );
    });
  });

  describe("addNewRepertoireTag", () => {
    it("should invoke the addRepertoireTag mutation", () => {
      const component = mountComponent();

      component.vm.addNewRepertoireTag(
        new RepertoireTag(
          Side.White,
          "",
          new RepertoirePosition("", "", Side.White),
          "",
          []
        ),
        "name"
      );

      expect(mutations.addRepertoireTag).toMatchSnapshot();
    });
  });

  describe("onScroll", () => {
    it("should go to the next position on scroll up", () => {
      const nextPosition = new RepertoirePosition("", "", Side.White);
      startPosition.children.push(new Move("", nextPosition));
      const component = mountComponent();
      component.vm.updateBoard = jest.fn();

      component.vm.onScroll({ deltaY: 1 } as any);

      expect(component.vm.updateBoard).toBeCalledWith(nextPosition);
    });

    it("should go to the next position on scroll down", () => {
      const nextPosition = new RepertoirePosition("", "", Side.White);
      startPosition.parents.push(nextPosition);
      const component = mountComponent();
      component.vm.updateBoard = jest.fn();

      component.vm.onScroll({ deltaY: -1 } as any);

      expect(component.vm.updateBoard).toBeCalledWith(nextPosition);
    });
  });
});
