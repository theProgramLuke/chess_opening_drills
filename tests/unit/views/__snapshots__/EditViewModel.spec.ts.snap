// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditViewModel addNewRepertoireTag should invoke the addRepertoireTag mutation 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blackRepertoire": Repertoire {
          "AddFromGame": [MockFunction],
          "AddMove": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveMove": [MockFunction],
          "RemoveRepertoireTag": [MockFunction],
          "positions": Array [],
          "tags": Array [],
        },
        "whiteRepertoire": Repertoire {
          "AddFromGame": [MockFunction],
          "AddMove": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveMove": [MockFunction],
          "RemoveRepertoireTag": [MockFunction],
          "positions": Array [],
          "tags": Array [
            RepertoireTag {
              "AddChild": [MockFunction],
              "AsSaved": [MockFunction],
              "RemoveChild": [MockFunction],
              "children": Array [],
              "fen": "",
              "forSide": 0,
              "id": "",
              "name": "",
              "position": RepertoirePosition {
                "AddChild": [MockFunction],
                "AddTrainingEvent": [MockFunction],
                "AsPgn": [MockFunction],
                "AsPgnMoveText": [MockFunction],
                "AsSaved": [MockFunction],
                "GetTurnLists": [MockFunction],
                "IncludeForTrainingMode": [MockFunction],
                "IsChildPosition": [MockFunction],
                "RootPaths": [MockFunction],
                "SideToMove": [MockFunction],
                "Unlink": [MockFunction],
                "VisitChildren": [MockFunction],
                "children": Array [],
                "comment": "",
                "easinessFactor": 0,
                "fen": "",
                "forSide": 0,
                "intervalIndex": 0,
                "myTurn": false,
                "nextRepetitionTimestamp": 0,
                "parents": Array [],
                "previousIntervalDays": 0,
                "trainingHistory": Array [],
              },
            },
          ],
        },
      },
      Object {
        "parent": RepertoireTag {
          "AddChild": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveChild": [MockFunction],
          "children": Array [],
          "fen": "",
          "forSide": 0,
          "id": "",
          "name": "",
          "position": RepertoirePosition {
            "AddChild": [MockFunction],
            "AddTrainingEvent": [MockFunction],
            "AsPgn": [MockFunction],
            "AsPgnMoveText": [MockFunction],
            "AsSaved": [MockFunction],
            "GetTurnLists": [MockFunction],
            "IncludeForTrainingMode": [MockFunction],
            "IsChildPosition": [MockFunction],
            "RootPaths": [MockFunction],
            "SideToMove": [MockFunction],
            "Unlink": [MockFunction],
            "VisitChildren": [MockFunction],
            "children": Array [],
            "comment": "",
            "easinessFactor": 0,
            "fen": "",
            "forSide": 0,
            "intervalIndex": 0,
            "myTurn": false,
            "nextRepetitionTimestamp": 0,
            "parents": Array [],
            "previousIntervalDays": 0,
            "trainingHistory": Array [],
          },
        },
        "tag": RepertoireTag {
          "AddChild": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveChild": [MockFunction],
          "children": Array [],
          "fen": "",
          "forSide": 0,
          "id": "",
          "name": "name",
          "position": RepertoirePosition {
            "AddChild": [MockFunction],
            "AddTrainingEvent": [MockFunction],
            "AsPgn": [MockFunction],
            "AsPgnMoveText": [MockFunction],
            "AsSaved": [MockFunction],
            "GetTurnLists": [MockFunction],
            "IncludeForTrainingMode": [MockFunction],
            "IsChildPosition": [MockFunction],
            "RootPaths": [MockFunction],
            "SideToMove": [MockFunction],
            "Unlink": [MockFunction],
            "VisitChildren": [MockFunction],
            "children": Array [],
            "comment": "",
            "easinessFactor": 0,
            "fen": "",
            "forSide": 0,
            "intervalIndex": 0,
            "myTurn": false,
            "nextRepetitionTimestamp": 0,
            "parents": Array [],
            "previousIntervalDays": 0,
            "trainingHistory": Array [],
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`EditViewModel onBoardMove should add a repertoire position with the new move 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blackRepertoire": Repertoire {
          "AddFromGame": [MockFunction],
          "AddMove": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveMove": [MockFunction],
          "RemoveRepertoireTag": [MockFunction],
          "positions": Array [],
          "tags": Array [],
        },
        "whiteRepertoire": Repertoire {
          "AddFromGame": [MockFunction],
          "AddMove": [MockFunction],
          "AsSaved": [MockFunction],
          "RemoveMove": [MockFunction],
          "RemoveRepertoireTag": [MockFunction],
          "positions": Array [],
          "tags": Array [
            RepertoireTag {
              "AddChild": [MockFunction],
              "AsSaved": [MockFunction],
              "RemoveChild": [MockFunction],
              "children": Array [],
              "fen": "",
              "forSide": 0,
              "id": "",
              "name": "",
              "position": RepertoirePosition {
                "AddChild": [MockFunction],
                "AddTrainingEvent": [MockFunction],
                "AsPgn": [MockFunction],
                "AsPgnMoveText": [MockFunction],
                "AsSaved": [MockFunction],
                "GetTurnLists": [MockFunction],
                "IncludeForTrainingMode": [MockFunction],
                "IsChildPosition": [MockFunction],
                "RootPaths": [MockFunction],
                "SideToMove": [MockFunction],
                "Unlink": [MockFunction],
                "VisitChildren": [MockFunction],
                "children": Array [],
                "comment": "",
                "easinessFactor": 0,
                "fen": "",
                "forSide": 0,
                "intervalIndex": 0,
                "myTurn": false,
                "nextRepetitionTimestamp": 0,
                "parents": Array [],
                "previousIntervalDays": 0,
                "trainingHistory": Array [],
              },
            },
          ],
        },
      },
      Object {
        "newMove": Move {
          "position": RepertoirePosition {
            "AddChild": [MockFunction],
            "AddTrainingEvent": [MockFunction],
            "AsPgn": [MockFunction],
            "AsPgnMoveText": [MockFunction],
            "AsSaved": [MockFunction],
            "GetTurnLists": [MockFunction],
            "IncludeForTrainingMode": [MockFunction],
            "IsChildPosition": [MockFunction],
            "RootPaths": [MockFunction],
            "SideToMove": [MockFunction],
            "Unlink": [MockFunction],
            "VisitChildren": [MockFunction],
            "children": Array [],
            "comment": "",
            "easinessFactor": 0,
            "fen": "new move fen",
            "forSide": 0,
            "intervalIndex": 0,
            "myTurn": false,
            "nextRepetitionTimestamp": 0,
            "parents": Array [],
            "previousIntervalDays": 0,
            "trainingHistory": Array [],
          },
          "san": "e5",
        },
        "parent": RepertoirePosition {
          "AddChild": [MockFunction],
          "AddTrainingEvent": [MockFunction],
          "AsPgn": [MockFunction],
          "AsPgnMoveText": [MockFunction],
          "AsSaved": [MockFunction],
          "GetTurnLists": [MockFunction],
          "IncludeForTrainingMode": [MockFunction],
          "IsChildPosition": [MockFunction],
          "RootPaths": [MockFunction],
          "SideToMove": [MockFunction],
          "Unlink": [MockFunction],
          "VisitChildren": [MockFunction],
          "children": Array [],
          "comment": "",
          "easinessFactor": 0,
          "fen": "",
          "forSide": 0,
          "intervalIndex": 0,
          "myTurn": false,
          "nextRepetitionTimestamp": 0,
          "parents": Array [],
          "previousIntervalDays": 0,
          "trainingHistory": Array [],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
